### PR TITLE
Notebook with intermediate level functions

### PR DESCRIPTION
**Pull request by AlinderS**
I have added a new notebook as extra content for lesson 1. The idea is that this notebook is available to students who feel they can benefit from it without further burdening the lecture. The notebook contains the points listed in #76 except for `zip` which is already covered in the manual for lesson 1, and I added lambda functions as well.
There is some room for improvement still. I don't have a solid example for lambda functions and there should maybe be something in the presentation to remind the teacher to mention the existence of the notebook to the students.

### CODE REVIEW

**Comment by AlinderS** [comment]
I asked the students in class today if they thought that something like this sounded useful to them. Basically everyone said yes.

File: 1-intro/intermediate.ipynb [hunk 1]
```diff
@@ -0,0 +1,354 @@
...
+    "with open(\"my_file\", \"w\") as f:\n",
+    "    # Your code here\n",
+    "    pass\n",
+    "\n",
+    "# Use instead\n",
+    "from pathlib import Path\n",
+    "\n",
+    "with Path(\"my_file\").open(\"w\") as f:\n",
+    "    # Your code here\n",
+    "    pass"
```
Comment (eerovaher) [inline 1-intro/intermediate.ipynb:224]: I don't think importing `pathlib` just to use `Path.open()` is more convenient than using the builtin `open()`. Much better examples would be something showcasing useful properties such as [`Path.parent`](https://docs.python.org/3/library/pathlib.html?highlight=pathlib#pathlib.PurePath.parent) or [`Path.name`](https://docs.python.org/3/library/pathlib.html?highlight=pathlib#pathlib.PurePath.name), or maybe something using [`Path.mkdir(parents=True, exist_ok=True)`](https://docs.python.org/3/library/pathlib.html?highlight=pathlib#pathlib.Path.mkdir) or [`Path.read_text()`](https://docs.python.org/3/library/pathlib.html?highlight=pathlib#pathlib.Path.read_text).

File: 1-intro/intermediate.ipynb [hunk 2]
```diff
@@ -0,0 +1,354 @@
...
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "27b44c04-c52c-4aaa-be43-035d00ce2f90",
+   "metadata": {},
+   "source": [
+    "### Sets\n",
```
Comment (eerovaher) [inline 1-intro/intermediate.ipynb:232]: Sets don't work very well with `numpy` arrays and I don't remember sets being useful in the projects. I don't think teaching about them is worth the effort.

File: 1-intro/intermediate.ipynb [hunk 3]
```diff
@@ -0,0 +1,354 @@
...
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "e5cca364-0576-4f2c-bb5e-93fb2040a705",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "add = lambda x, y: x + y\n",
```
Comment (eerovaher) [inline 1-intro/intermediate.ipynb:302]: Assigning lambda functions to variables is a code smell, see [PEP 8](https://peps.python.org/pep-0008/#programming-recommendations) and [`Ruff rule E731`](https://docs.astral.sh/ruff/rules/lambda-assignment/).

File: 1-intro/intermediate.ipynb [hunk 4]
```diff
@@ -0,0 +1,354 @@
...
+   "outputs": [],
+   "source": [
+    "# Generate some data that you don't know what it will be beforehand\n",
+    "import numpy as np\n",
+    "rng = np.random.default_rng()\n",
+    "multipliers = 10 * rng.random(rng.integers(3, 10, 1))\n",
+    "print(multipliers)\n",
+    "\n",
+    "# Do some operation on the data that is not super simple\n",
+    "complicated_numbers = [(lambda x: (x + 1) / len(multipliers))(num) for num in multipliers]\n",
```
Comment (eerovaher) [inline 1-intro/intermediate.ipynb:328]: The following is exactly equivalent, but much simpler code:
```suggestion
    "complicated_numbers = ((multipliers + 1) / len(multipliers)).tolist()\n",
```
Note that in practice the `.tolist()` is usually not needed, which would make the code even shorted.

Good examples of using lambda functions would be [keys for `sorted()`](https://docs.python.org/3/howto/sorting.html#key-functions). For example, suppose we have some numbers and we want to sort them by their digit sum:
```python
>>> numbers = [19, 100, 123]
>>> sorted_by_digit_sum = sorted(numbers, key=lambda x: sum(int(d) for d in str(x)))
>>> print(sorted_by_digit_sum)
[100, 123, 19]
```

**Review by eerovaher — COMMENTED** [review]
`map()`, `filter()` and sets are not useful for the projects, so I don't think teaching them is worth the effort. Most of the other ideas are useful, but I would rather add or emphasize exercises that would make use of those ideas. `pathlib` is only mentioned in a presentation, so it would be good to introduce it more thoroughly in a manual, but I don't see why it would be better to create a brand new notebook instead of updating an existing one.

File: 1-intro/intermediate.ipynb [hunk 5]
```diff
@@ -0,0 +1,354 @@
...
+    "# Intermediate Python functions and tips\n",
+    "This notebook contains a list of functions and tips that are not required to write working Python code, and therefore usually skipped when teaching the basics, but are useful or convenient."
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "fe1dba59-aaaf-4312-94c8-61f4aede26b4",
+   "metadata": {},
+   "source": [
+    "### Enumerate\n",
```
[the remark this answers is not in the document]
Comment (AlinderS) [reply 1-intro/intermediate.ipynb:17]: Yeah, I think that putting an explanation of how `enumerate()` works in manual 1 where it is used makes sense.

File: 1-intro/intermediate.ipynb [hunk 6]
```diff
@@ -0,0 +1,354 @@
...
+    "        yield num\n",
+    "        num += 1"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "381ce226-2e74-4813-9f12-26d96b8480ce",
+   "metadata": {},
+   "source": [
+    "### The unpacking operator\n",
```
[the remark this answers is not in the document]
Comment (AlinderS) [reply 1-intro/intermediate.ipynb:111]: The unpacking operator is used in a solution but it is never explained, and if we don't want to devote more real estate in the manual to it, then it belongs here.

File: 1-intro/intermediate.ipynb (hunk 1 above)
Comment (AlinderS) [reply 1-intro/intermediate.ipynb:224]: [`Ruff rule PTH123`](https://docs.astral.sh/ruff/rules/builtin-open/) says to use `Path.open()` instead of `open()`.
Including some of those other methods here seems like a good idea to me.

File: 1-intro/intermediate.ipynb (hunk 4 above)
Comment (AlinderS) [reply 1-intro/intermediate.ipynb:328]: I have a hard time coming up with good examples for lambda functions, and I very rarely use them myself. Is it maybe best to drop them completely?

File: 1-intro/intermediate.ipynb (hunk 2 above)
Comment (AlinderS) [reply 1-intro/intermediate.ipynb:232]: A set is used in a solution in session 1. I think it is good to explain how they work a bit more.
https://github.com/Lund-Observatory-Teaching/lundpython/blob/87d12f4f430f804263d5dbe6f2d011be50bec9a7/1-intro/exercises.ipynb#L181C1-L181C48
Adding a caveat about `numpy` is probably a good idea.

File: 1-intro/intermediate.ipynb (hunk 6 above)
[the remark this answers is not in the document]
Comment (eerovaher) [reply 1-intro/intermediate.ipynb:111]: Iterable unpacking is not useful for the courses, so I wouldn't spend any time on it. The one-liner exercise solution can be replaced with an even shorter one-liner:
```python
print("\n".join(min(i, 10 - i) * "*" for i in range(1, 10)))
```
The closely related concept of accepting and passing on `**kwargs` can be useful for writing plotting functions because it allows passing on arguments to the underlying `matplotlib` function without having to write a very verbose function signature. Such use of `**kwargs` is demonstrated in the Dynamical Astronomy notebooks in a clear example. The reasoning for having it there and not in the Python introduction materials was that spreading the lessons out in time makes them more accessible and that most of the students taking the Python introduction will also take Dynamical Astronomy anyways. Using `**kwargs` could be brought up in the `matplotlib` lesson, if there is time for it.

File: 1-intro/intermediate.ipynb (hunk 1 above)
Comment (eerovaher) [reply 1-intro/intermediate.ipynb:224]: My point was that if I had code that was not using `pathlib` already, then using `Path.open()` instead of the built-in `open()` would be _less_ convenient because it would require an additional import. Preferring `Path.open()` only makes sense if the code is using `pathlib` anyways.

File: 1-intro/intermediate.ipynb (hunk 2 above)
Comment (eerovaher) [reply 1-intro/intermediate.ipynb:232]: Sets are not useful for the courses so it would be better to avoid using it in the exercise solution, or to change the exercise entirely.

File: 1-intro/intermediate.ipynb (hunk 4 above)
Comment (eerovaher) [reply 1-intro/intermediate.ipynb:328]: Lambda functions are syntactic sugar for passing a single-expression function to some function that accepts other functions as arguments if there is no suitable function defined already. It is always possible to define the function with `def` and then pass it on by name instead of defining it in-place with lambda.

I use lambda functions for [monkeypatching functions in unit tests](https://docs.pytest.org/en/stable/how-to/monkeypatch.html#monkeypatching-functions), but you can see that the `pytest` documentation is using `def` instead. 

In more functionally-oriented languages lambda functions would often be used with `map()` and `filter()` (or with whatever would be equivalent), but in Python using comprehensions is more convenient.

In summary, lambda functions are not useful for the courses, I wouldn't spend time on them.

Comment (AlinderS) [reply 1-intro/intermediate.ipynb:328]: Ok, agreed. I will remove the section on lambda functions.
